sdk/log/DESIGN.md

```diff
@@ -0,0 +1,170 @@
+are defined as follows:
+
+```go
+func WithAttributeCountLimit(limit int) Option
+
+func WithAttributeValueLengthLimit(limit int) Option
+```
+
+### LogRecordProcessor and LogRecordExporter  
+
+Both [LogRecordProcessor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordprocessor)
+and [LogRecordExporter](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordexporter)
+are defined via an `Exporter` interface:
+
+```go
+func WithExporter(Exporter) Option {
+	return nil
+}
+
+type Exporter interface {
+	Export(ctx context.Context, records []*Record) error
+	Shutdown(ctx context.Context) error
+	ForceFlush(ctx context.Context) error
+}
+```
+
+The `Record` struct represents the [ReadWriteLogRecord](https://opentelemetry.io/docs/specs/otel/logs/sdk/#readwritelogrecord).
+
+```go
+type Record struct { /* ... */ }
+
+func (r *Record) Timestamp()
+
+func (r *Record) SetTimestamp(t time.Time)
+
+func (r *Record) ObservedTimestamp() time.Time
+
+func (r *Record) SetObservedTimestamp(t time.Time)
+
+func (r *Record) Severity() log.Severity
+
+func (r *Record) SetSeverity(level log.Severity)
+
+func (r *Record) SeverityText() string
+
+func (r *Record) SetSeverityText(text string)
+
+func (r *Record) Body() log.Value
+
+func (r *Record) SetBody(v log.Value)
+
+func (r *Record) WalkAttributes(f func(log.KeyValue) bool)
+
+// SetAttributes sets and overrides the attributes of the log record.
+func (r *Record) SetAttributes(attrs ...log.KeyValue)
+
+func (r *Record) TraceID() trace.TraceID
+
+func (r *Record) SpanID() trace.SpanID
+
+func (r *Record) TraceFlags() trace.TraceFlags
+
+func (r *Record) Resource() resource.Resource
+
+func (r *Record) InstrumentationScope() instrumentation.Scope
+
+func (r *Record) AttributeValueLengthLimit() int
+
+func (r *Record) AttributeCountLimit() int
+```
+
+The user can implement a custom [LogRecordProcessor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logrecordprocessor)
+by implementing a `Exporter` decorator.
+
+This is similar to the design of HTTP server middleware
+which is a wrapper of `http.Handler`.[^2]
+
+[Simple processor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#simple-processor)
+is a achieved by simply passing a bare-exporter.
+
+[Batching processor](https://opentelemetry.io/docs/specs/otel/logs/sdk/#batching-processor)
+is a achieved by wrapping an expoter with `Batcher`:
+
+```go
+type Batcher struct { /* ... */ }
+
+func NewBatchingExporter(exporter Exporter, opts ...BatchingOption) *Batcher
+
+func (b *Batcher) Export(ctx context.Context, records []*Record)
+
+func (b *Batcher) Shutdown(ctx context.Context) error
+
+func (b *Batcher) ForceFlush(ctx context.Context) error
+
+type BatchingOption interface { /* ... */ }
+
+func WithInterval(d time.Duration) BatchingOption
+
+func WithTimeout(d time.Duration) BatchingOption
+
+func WithQueueSize(max int) BatchingOption
+
+func WithBatchSize(max int) BatchingOption
```

Comment:
PTAL https://github.com/open-telemetry/opentelemetry-go/pull/4954/commits/3cb2da401efab3d55089f870e6049f4fbc237ed7

I decided to rename as following:
- `WithMaxQueueSize` as suggested
- `WithExportInterval` - I think that adding max/min can be misleading. The spec says that the interval is as specified "the delay interval in milliseconds between two consecutive exports". I would rather say that this is maximum interval as someone can execute `ForceFlush` or `Shutdown` before the interval elapses. However, I think it is better to stick with simply `WithExportInterval`.
- `WithExportTimeout` as suggested
- `WithExportMaxBatchSize` instead of `WithMaxExportBatchSize` as I think having `WithExport` prefix for options affecting the export is nice